src/protocol/step.rs

```diff
@@ -74,9 +74,14 @@ impl<S: Step + ?Sized> StepNarrow<S> for Descriptive {
             assert!(!s.contains('/'), "The string for a step cannot contain '/'");
         }
 
-        Self {
-            id: self.id.clone() + "/" + step.as_ref(),
+        let mut id = self.id.clone() + "/";
+        #[cfg(feature = "step-trace")]
+        {
+            id += &(std::any::type_name::<S>().to_owned() + "::");
         }
+        id += step.as_ref();
```

Comment:
Did you think about the challenge of recording the enum variant as well, rather than the `AsRef<str>` output?

Reply:
Yes, initially that was my preferred option and we could use [strum](https://docs.rs/strum/latest/strum/) for that. One challenge would be the `struct` steps (i.e., `struct BitOpStep(usize)`), and tuple like enum variants with `repeat64str!` macro. (i.e., `IpaProtocolStep::Sort(usize)`).

I think we can do something similar to thiserror::Error derive macro where we annotate the step keys and values. With that approach we can generate both Descriptive (as_ref()) and Compact (ints) gates.

So I think we should consider thiserror::Error kind of approach. Since that'll require more consideration, use this simpler `AsRef<str>` output approach for the time being. I think the codegen algorithm shouldn't change that much with either approach.

Reply:
strum IntoStaticStr seems idea for this.  Want to add that to this patch so that you can get the entire thing together?

Reply:
I think I'll make a separate diff for that because we'll need to deal with `struct` steps. I guess the easiest solution is to convert `struct BitOpStep(usize)` to `enum Step { BitOp(usize) }`. For this, proc macro is also useful to enforce all step definitions must be of enum type.